File: src/api/ipc/agent.ts
import { invoke, isTauri } from '@tauri-apps/api/core'
import type {
  AgentSettings,
  AgentProfileInput,
  AgentModelInput,
} from '@/modules/ai/settings'
import type {
  AgentAttachment,
  AgentContext,
  AgentConversationSummary,
  AgentRun,
} from '@/modules/ai/types'

export interface StartInput {
  profileId: string
  prompt: string
  context: AgentContext
  conversationId?: string
  attachments: AgentAttachment[]
}
export interface SendInput {
  runId: string
  prompt: string
  context: AgentContext
  attachments: AgentAttachment[]
}
function call<T>(command: string, args?: Record<string, unknown>): Promise<T> {
  if (!isTauri())
    return Promise.reject(new Error('请在 MiraiNote 桌面程序中配置和使用 AI'))
  return invoke<T>(command, args)
}
export function agentError(error: unknown): string {
  if (error instanceof Error) return error.message
  if (
    typeof error === 'object' &&
    error &&
    'message' in error &&
    typeof error.message === 'string'
  )
    return error.message
  return 'AI 操作失败，请重试'
}
export const agentApi = {
  isDesktop: isTauri,
  getSettings: () =>
    isTauri()
      ? call<AgentSettings>('agent_get_settings')
      : Promise.resolve({ activeId: '', profiles: [] } as AgentSettings),
  save: (input: AgentProfileInput, clearKey: boolean) =>
    call<AgentSettings>('agent_save_profile', { input, clearKey }),
  remove: (id: string) => call<AgentSettings>('agent_delete_profile', { id }),
  activate: (id: string) =>
    call<AgentSettings>('agent_activate_profile', { id }),
  revealKey: (id: string) => call<string>('agent_reveal_key', { id }),
  models: (input: AgentModelInput) =>
    call<string[]>('agent_list_models', { input }),
  test: (input: AgentProfileInput, clearKey: boolean) =>
    call<string>('agent_test_profile', { input, clearKey }),
  start: (input: StartInput) => call<AgentRun>('agent_start', { input }),
  send: (input: SendInput) => call<AgentRun>('agent_send', { input }),
  step: (runId: string) => call<AgentRun>('agent_step', { runId }),
  cancel: (runId: string) => call<void>('agent_cancel', { runId }),
  forget: (runId: string) => call<void>('agent_forget', { runId }),
  conversations: (documentId: string) =>
    isTauri()
      ? call<AgentConversationSummary[]>('agent_list_conversations', {
          documentId,
        })
      : Promise.resolve([]),
  openConversation: (documentId: string, conversationId: string) =>
    call<AgentRun>('agent_open_conversation', { documentId, conversationId }),
  renameConversation: (
    documentId: string,
    conversationId: string,
    title: string
  ) =>
    call<AgentConversationSummary>('agent_rename_conversation', {
      documentId,
      conversationId,
      title,
    }),
  deleteConversation: (documentId: string, conversationId: string) =>
    call<void>('agent_delete_conversation', { documentId, conversationId }),
}

File: src-tauri/src/commands/agent.rs
use crate::{
    error::AppResult,
    services::agent::{
        config::{ProfileInput, PublicSettings},
        history::Summary,
        manager::{AgentService, SendInput, Snapshot, StartInput},
        models::ModelListInput,
    },
};
use tauri::{AppHandle, State};

#[tauri::command]
pub async fn agent_get_settings(
    app: AppHandle,
    service: State<'_, AgentService>,
) -> AppResult<PublicSettings> {
    service.settings(&app).await
}
#[tauri::command]
pub async fn agent_save_profile(
    app: AppHandle,
    service: State<'_, AgentService>,
    input: ProfileInput,
    clear_key: bool,
) -> AppResult<PublicSettings> {
    service.save(&app, input, clear_key).await
}
#[tauri::command]
pub async fn agent_delete_profile(
    app: AppHandle,
    service: State<'_, AgentService>,
    id: String,
) -> AppResult<PublicSettings> {
    service.remove(&app, &id).await
}
#[tauri::command]
pub async fn agent_activate_profile(
    app: AppHandle,
    service: State<'_, AgentService>,
    id: String,
) -> AppResult<PublicSettings> {
    service.activate(&app, &id).await
}
#[tauri::command]
pub async fn agent_reveal_key(
    app: AppHandle,
    service: State<'_, AgentService>,
    id: String,
) -> AppResult<String> {
    service.reveal_key(&app, &id).await
}
#[tauri::command]
pub async fn agent_list_models(
    app: AppHandle,
    service: State<'_, AgentService>,
    input: ModelListInput,
) -> AppResult<Vec<String>> {
    service.list_models(&app, input).await
}
#[tauri::command]
pub async fn agent_test_profile(
    app: AppHandle,
    service: State<'_, AgentService>,
    input: ProfileInput,
    clear_key: bool,
) -> AppResult<String> {
    service.test(&app, input, clear_key).await
}
#[tauri::command]
pub async fn agent_start(
    app: AppHandle,
    service: State<'_, AgentService>,
    input: StartInput,
) -> AppResult<Snapshot> {
    service.start(&app, input).await
}
#[tauri::command]
pub async fn agent_send(
    app: AppHandle,
    service: State<'_, AgentService>,
    input: SendInput,
) -> AppResult<Snapshot> {
    service.send(&app, input).await
}
#[tauri::command]
pub async fn agent_step(
    app: AppHandle,
    service: State<'_, AgentService>,
    run_id: String,
) -> AppResult<Snapshot> {
    service.step(&app, &run_id).await
}
#[tauri::command]
pub async fn agent_cancel(
    app: AppHandle,
    service: State<'_, AgentService>,
    run_id: String,
) -> AppResult<()> {
    service.cancel(&app, Some(&run_id)).await
}
#[tauri::command]
pub async fn agent_forget(
    app: AppHandle,
    service: State<'_, AgentService>,
    run_id: String,
) -> AppResult<()> {
    service.forget(&app, &run_id).await
}
#[tauri::command]
pub async fn agent_list_conversations(
    app: AppHandle,
    service: State<'_, AgentService>,
    document_id: String,
) -> AppResult<Vec<Summary>> {
    service.list_conversations(&app, &document_id).await
}
#[tauri::command]
pub async fn agent_open_conversation(
    app: AppHandle,
    service: State<'_, AgentService>,
    document_id: String,
    conversation_id: String,
) -> AppResult<Snapshot> {
    service
        .open_conversation(&app, &document_id, &conversation_id)
        .await
}
#[tauri::command]
pub async fn agent_rename_conversation(
    app: AppHandle,
    service: State<'_, AgentService>,
    document_id: String,
    conversation_id: String,
    title: String,
) -> AppResult<Summary> {
    service
        .rename_conversation(&app, &document_id, &conversation_id, &title)
        .await
}
#[tauri::command]
pub async fn agent_delete_conversation(
    app: AppHandle,
    service: State<'_, AgentService>,
    document_id: String,
    conversation_id: String,
) -> AppResult<()> {
    service
        .delete_conversation(&app, &document_id, &conversation_id)
        .await
}
